Automattic/jetpack
# Jetpack Search/Sync: ensure syncing of post meta and taxonomies that are indexed (#2)

Our search index includes a fixed list of post meta keys and custom taxonomies in order to limit the number of fields in the index (see documentation https://jetpack.com/support/search/frequently-asked-questions/#troubleshoot-custom-tax).

We have so far not been syncing the list from wp.com into Jetpack which has had many downsides:
- Uncertainty if it worked in all cases (also lack of tests)
- No easy way to add new fields
- No visibility into what is indexed
- No way to double check that the config that the user has will work.
- No way to provide post meta faceting in search

This fixes all of that.

The class-search.php file will be replacing the old-style class.jetpack-sync-module-search.php that we are currently using on wp.com for the indexing. See https://github.com/Automattic/wpes-lib/blob/master/src/indices/jetpack-search-v2-index.php#L842

Fixes #15672
Fixes #8663

See #18870 for historical discussions.
See also pcNPJE-5Y-p2.

#### Does this pull request change what data or activity we track or use?

This auto expands the post meta that is synced for sites with Jetpack Instant Search enabled. This data can already be synced by users if they manually add it to allow lists. 

#### Testing instructions:

New unit tests are included. The unit tests are one of the more important pieces since I expect that the list of post meta and taxonomies will evolve over time and we want to make sure that it is trivial to add a new field.

Manual Instructions:

Spin up a new Jetpack powered site
Add a compatible Jetpack Search Plan
add search allowed meta to a post ( jetpack-search-meta0, entity_types, minimum_amount, etc )
Verify the Sync Queue is empty (data synced) 
Verify the meta is synced to the WP.com Cache

#### Proposed changelog entry for your changes:
* Search: adds explicit post meta keys and taxonomies to sync into the index. Particularly important for integrating with third-party plugins.
